js/lib/BTCUtils.js

```diff
@@ -0,0 +1,659 @@
+   */
+  extractOutpoint: (input) => {
+    return utils.safeSlice(input, 0, 36);
+  },
+
+  /**
+   * @notice                Extracts the outpoint tx id from an input
+   * @dev                   32 byte tx id
+   * @param {Uint8Array}    input The input
+   * @returns {Uint8Array}  The tx id (little-endian bytes)
+   */
+  // TODO: no test, check against function that uses this
+  extractInputTxIdLE: (input) => {
+    return utils.safeSlice(input, 0, 32);
+  },
+
+  /**
+   * @notice                Extracts the outpoint index from an input
+   * @dev                   32 byte tx id
+   * @param {Uint8Array}    input The input
+   * @returns {Uint8Array}  The tx id (big-endian bytes)
+   */
+  // TODO: no test, check against function that uses this
+  extractInputTxId: (input) => {
+    let leId = module.exports.extractInputTxIdLE(input);
+    return module.exports.reverseEndianness(leId);
+  },
+
+  /**
+   * @notice                Extracts the LE tx input index from the input in a tx
+   * @dev                   4 byte tx index
+   * @param {Uint8Array}    input The input
+   * @returns {Uint8Array}  The tx index (little-endian bytes)
+   */
+  // TODO: no test, check against function that uses this
+  extractTxIndexLE: (input) => {
+    return utils.safeSlice(input, 32, 36);
+  },
+
+  /**
+   * @notice                Extracts the LE tx input index from the input in a tx
+   * @dev                   4 byte tx index
+   * @param {Uint8Array}    input The input
+   * @returns {BigInt}      The tx index (big-endian uint)
+   */
+  // TODO: no test, check against function that uses this
+  extractTxIndex: (input) => {
+    let leIndex = module.exports.extractTxIndexLE(input);
+    let beIndex = module.exports.reverseEndianness(leIndex);
+    return utils.bytesToUint(beIndex);
+  },
+
+  /* ****** */
+  /* Output */
+  /* ****** */
+
+  /**
+   * @notice                Determines the length of an output
+   * @dev                   5 types: WPKH, WSH, PKH, SH, and OP_RETURN
+   * @param {Uint8Array}    output The output
+   * @returns {number}      The length indicated by the prefix, error if invalid length
+   */
+  determineOutputLength: (output) => {
+    let len = utils.safeSlice(output, 8, 9)[0];
+
+    if (len > 0xfd) {
+      throw new Error("Multi-byte VarInts not supported");
+    }
+
+    return BigInt(len) + 8n + 1n; // 8 byte value, 1 byte for len itself
+  },
+
+  /**
+   * @notice                Extracts the output at a given index in the TxIns vector
+   * @dev                   Iterates over the vout. If you need to extract multiple, write a custom function
+   * @param {Uint8Array}    vout The _vout to extract from
+   * @param {number}        index The 0-indexed location of the output to extract
+   * @returns {Uint8Array}  The specified output
+   */
+  extractOutputAtIndex: (vout, index) => {
+    let len;
+    let remaining;
+    let offset = 1n;
```

Comment:
Can we move `let len` and `let remaining` into the loop

Reply:
remaining can be moved inside, but `len` needs to be outside so it can be referenced in the return statement.